darwin: add battery status (WIP)

juanarbol opened the pull request:
This is WIP with the intention of doing corrections if needed and may actually define the `uv_battery_info_t` struct and refine the implementation to handle all that the `getBattery` needs.

Refs: https://github.com/libuv/libuv/issues/4172

bnoordhuis commented on include/uv.h:
```diff
@@ -1160,6 +1162,10 @@ struct uv_cpu_info_s {
 };
 
+struct uv_battery_info_s {
+  int charging;
```
Ambiguous name, can be interpreted as a boolean. Consider naming it e.g. `charge_time` or `charge_time_in_secs`.

Also, since it can't go negative, it's probably better to make it an `unsigned int`. Maybe make it a `uint64_t` out of an abundance of caution.

You probably want to add `void *reserved[7];` so there's space to add more features, because, once it's released, the size of the struct can never, ever change again.

anonrig replied:
> Because, once it's released, the size of the struct can never, ever change again.

@bnoordhuis Why can't the size of the struct can never change?

bnoordhuis replied:
ABI stability.

bnoordhuis commented on include/uv.h:
```diff
@@ -1295,6 +1301,7 @@ enum {
 UV_EXTERN int uv_thread_getpriority(uv_thread_t tid, int* priority);
 UV_EXTERN int uv_thread_setpriority(uv_thread_t tid, int priority);
 
+UV_EXTERN int uv_battery_info(uv_battery_info_t* battery_info);
```
Can you document this in docs/src/misc.rst?

juanarbol replied:
Will do!

bnoordhuis commented on src/unix/darwin-battery.c:
```diff
@@ -0,0 +1,66 @@
+  int err;
+
+  err = UV_ENOENT;
+  iokit_framekwork_handle = dlopen("/System/Library/Frameworks/"
+                                  "IOKit.framework/IOKit",
+                                  RTLD_LAZY | RTLD_LOCAL);
```
```suggestion
  iokit_framekwork_handle = dlopen("/System/Library/Frameworks/"
                                   "IOKit.framework/IOKit",
                                   RTLD_LAZY | RTLD_LOCAL);
```

bnoordhuis commented on src/unix/darwin-battery.c:
```diff
@@ -0,0 +1,66 @@
+    goto out;
+
+  err = UV_EINVAL;
+  info->charging = pIOPSGetTimeRemainingEstimate();
```
You should assign this to a local variable first because implementation details like the -1 and -2 status codes shouldn't leak out to callers.

juanarbol replied:
On it!

How are we going to handle -1 and -2 states? As long as the struct will store only positive numbers?

oluan replied:
I've added further members to the structure, but I think the question remains: how to handle platform-specific values like -1, or BYTE(255) in Win32 case.

https://github.com/libuv/libuv/pull/4271

bnoordhuis commented on test/test-platform-output.c:
```diff
@@ -197,5 +198,12 @@ TEST_IMPL(platform_output) {
   printf("  version: %s\n", uname.version);
   printf("  machine: %s\n", uname.machine);
 
+#if defined(__APPLE__)
+  err = uv_battery_info(&battery);
+  printf("uv_battery_info:\n");
+  printf("  charging: %d\n", battery.charging);
+  ASSERT_OK(err);
+#endif
```
```suggestion
  err = uv_battery_info(&battery);
  printf("uv_battery_info:\n");
  printf("  charging: %d\n", battery.charging);
  ASSERT_OK(err);
```
And add a stub for other platforms. The stub should probably return UV_ENOSYS, to indicate it's unimplemented.

juanarbol replied:
WIll do!

juanarbol replied:
Will change the assertion to this, in the meantime `ASSERT_OK(err && err == UV_ENOSYS);`

juanarbol replied:
I moved this file into the `unix` folder and renamed it just `battery.c` Using macros we can populate the struct, so we just handle two files, one for UNIX, and another one for Win.

bnoordhuis reviewed (commented):
Thanks for the PR. Left some comments.

bnoordhuis commented:
@juanarbol is this ready for re-review? Some of the comments don't seem to be addressed.

juanarbol commented:
I had to spin a Linux VM. I'm working on the Linux patch. This is WIP.

juanarbol commented:
Some IOKit calls or "core foundation" are unavailable for the macOS machine, but I am not sure...

juanarbol commented:
This is now ready for review, but I have a question about how this should behave.

For now, if there is no battery for both Linux and Darwin, it will return `UV_EPERM`, and nothing else will be done.

In the Linux implementation, if it fails to read any other "battery information" file, I will return `UV_EPERM`, but macOS will leave the struct with garbage in the struct members and will return 0.

What do reviewers think is a better behavior to adopt? I propose two. I'm open to modifying this to a better one or unifying this into a decided one (and make the CI pass, for sure)

cjihrig commented on src/unix/battery.c:
```diff
@@ -0,0 +1,252 @@
+#if defined(__APPLE__) || TARGET_OS_IPHONE
+#include <dlfcn.h>
+#include "darwin-stub.h"
+#endif // defined(__APPLE__) || TARGET_OS_IPHONE
```
```suggestion
#endif /* defined(__APPLE__) || TARGET_OS_IPHONE */
```

cjihrig replied:
I'll stop commenting on them individually, but can you change all of the comments in this PR to C style.

cjihrig commented on src/win/battery.c:
```diff
@@ -0,0 +1,26 @@
+#include "uv.h"
+
+int uv_battery_info(uv_battery_info_t* info) {
+  // To be implemented
```
```suggestion
  /* To be implemented. */
```

cjihrig commented on test/test-platform-output.c:
```diff
@@ -197,5 +198,18 @@ TEST_IMPL(platform_output) {
+  printf("  discharge_time_in_secs: %lld\n", battery.discharge_time_in_secs);
+  printf("  charge_time_in_secs: %lld\n", battery.charge_time_in_secs);
+#if !defined(_WIN32)
+  // In desktops the battery is not present.
```
```suggestion
  /* In desktops the battery is not present. */
```

cjihrig commented on src/unix/battery.c:
```diff
@@ -0,0 +1,252 @@
+#include "darwin-stub.h"
+#endif // defined(__APPLE__) || TARGET_OS_IPHONE
+
+// macOS won't need any of this
```
```suggestion
/* macOS won't need any of this. */
```

This comment probably isn't really needed though.

cjihrig commented on src/unix/battery.c:
```diff
@@ -0,0 +1,252 @@
+  FILE* fp;
+
+  err = UV_EPERM;
+  fp = uv__open_file("/sys/class/power_supply/BAT0");
```
Would `uv__slurp()` help with all of this file reading logic?

bnoordhuis replied:
Another thing: there can be multiple batteries. Linux's ACPI smart battery driver supports up to 4 (BAT0 to BAT3.)

I suppose you should try them all and merge the results? Either that or the documentation should make explicit that it only queries the first one.

cjihrig commented on src/unix/battery.c:
```diff
@@ -0,0 +1,252 @@
+
+  // If there are no sources, we're on a desktop or server, nothin' to do
+  if (pCFArrayGetCount(sources) == 0) {
+    err = UV_EPERM;
```
`EPERM` feels like the wrong error code. Would `ENOTSUP` make more sense?

bnoordhuis commented on src/unix/battery.c:
```diff
@@ -0,0 +1,252 @@
+                                            const char*,
+                                            CFStringEncoding);
+
+  CFBooleanRef isChargingValue;
+  CFDictionaryRef powerSource;
+  CFNumberRef currentCapacity;
+  CFNumberRef timeRemainingValue;
```
Libuv uses snake_case for locals, ex. `is_charging_value`, `power_source`, etc.

(`pCFArrayGetCount` and friends get a pass since they mirror Apple's naming convention.)

bnoordhuis commented on src/unix/battery.c:
```diff
@@ -0,0 +1,252 @@
+    pCFDictionaryGetValue(powerSource,
+        is_charging ? S("Time to Full Charge") : S("Time to Empty"));
+  if (timeRemainingValue != NULL) {
+    pCFNumberGetValue(timeRemainingValue, pkCFNumberIntType, &holder);
```
Should maybe be called `current_capacity_value` or something like that? Unclear what `holder` refers to.

juanarbol replied:
I just created **one** variable to hold/store all values, I just rewrite that holder during the whole process, I will create all the required variables for this :)

bnoordhuis replied:
It's okay to have a single variable that's used repeatedly but we usually give those a single-letter (or very short) name, like `v`.

bnoordhuis commented on src/unix/battery.c:
```diff
@@ -0,0 +1,252 @@
+    info->charge_time_in_secs = 0;
+  }
+out:
+  if (iokit_framekwork_handle != NULL) {
+    dlclose(iokit_framekwork_handle);
+  }
```
Here and elsewhere:
```suggestion
  if (iokit_framekwork_handle != NULL)
    dlclose(iokit_framekwork_handle);
```

bnoordhuis commented on src/unix/battery.c:
```diff
@@ -0,0 +1,252 @@
+
+  if (strstr(buf, "Charging") != NULL) {
+    info->is_charging = 1;
+    fp = uv__open_file("/sys/class/power_supply/BAT0/time_to_full_now");
```
I'm fairly sure time_to_full_now and time_to_empty_now don't exist.

I'm not sure how you'd estimate time-to-full, I don't think the kernel exposes that directly. You're probably supposed to piece it together from charge_full and charge_now, or something like that.

juanarbol replied:
I think it does https://github.com/torvalds/linux/blob/master/drivers/power/supply/power_supply_sysfs.c#L209

I validated that with some coworker's machine.

bnoordhuis replied:
Ah, okay. Looks like there are only two or three drivers that implement it. The files don't exist on my system.

bnoordhuis commented:
> In the Linux implementation, if it fails to read any other "battery information" file, I will return UV_EPERM, but macOS will leave the struct with garbage in the struct members and will return 0.

If you return an error, it's okay to leave the struct fields untouched. If you return 0 (no error), they must be set to something sensible (like zero.)

juanarbol reviewed (commented):
Thanks for taking the time reviewing this, I will address all your suggestions as soon as I can :)

juanarbol commented:
I am still working on this; this is not stale; I am just not having the best time w/out a Linux machine with a better attached to it.

juanarbol commented:
> I suppose you should try them all and merge the results? Either that or the documentation should make explicit that it only queries the first one.

I have the loop counting and reading all batteries inside this code (validated with my coworker's machine), but I'm not sure what do you mean with "merge" results, what if one of the three batteries is charging but the rest are not? I was considering using a `uv_cpu_info_t` array, likes `uv_cpu_info()`, return an array of batteries instead. But I think windows, nor darwin support multiple batteries.

>That said, it's probably better to move the platform-specific code into linux.c, darwin.c, etc., and have a UV_ENOSYS stub for platforms where it's unimplemented. (https://github.com/libuv/libuv/pull/4258#discussion_r1460862692)

I put the conditional inclusion due to, macOS way won't need stdlib, just link to IOKit. Does that worth move split all this into separate files? (I'm genuinely asking, I don't know)

juanarbol commented:
ping @bnoordhuis

calvin2021y commented:
Is android/ios supported ?

juanarbol commented:
> Is android/ios supported ?

I think IOKit gives immediate support to iOS, not quite sure about android.

juanarbol commented:
I'm sorry for the delay, this new revision is rebased on top v1.x, it follows all the desired/discussed behavior, except for multiple batteries support as I could not find any good mechanism to merge all the results.

Three batteries could be charging and one is not (for example), my initial approach was creating something like `uv_cpu_info` but macOS and windows won't like that extra memory space allocated for nothing (AFAIK they support only one battery).

I think that multiple batteries support is something we can achieve in the short future.

BTW, the windows implementation is being addressed in https://github.com/libuv/libuv/pull/4271

bnoordhuis commented:
@juanarbol now _I_ am sorry for the delay :) (was on holiday in August)

There are still some unresolved-looking comments. If they are in fact resolved, can you mark them as such (and if not, address them)?

juanarbol commented:
@bnoordhuis no worries! Thanks for reviewing this again.

This new patch addresses all the comments. I'm not the biggest fan of including `inttypes` in the test, but avoiding it will create a warning in clang or gnu in case of fixing it for clang.

This patch is very intrusive due to stub the non-supported (yet) platforms. See

> That said, it's probably better to move the platform-specific code into linux.c, darwin.c, etc., and have a UV_ENOSYS stub for platforms where it's unimplemented.

https://github.com/libuv/libuv/pull/4258#discussion_r1460862692

juanarbol commented:
@bnoordhuis I came up w/ a bit more resilient approach for linux, everything should be in `uevent`, the device model will eventually create that file.

I've read couple bits of the kernel, normally BAT0 is the main battery in a laptop, not sure if we need to handle more than that, the other references seems to be associated to UPCs and stuff.

Luckily, everything we're reading seems to be part of the ACPI standard.
